src/node/mini_miner.cpp

```diff
@@ -0,0 +1,265 @@
+        const CFeeRate b_feerate(b->second.GetModFeesWithAncestors(), b->second.GetSizeWithAncestors());
+        if (a_feerate != b_feerate) {
+            return a_feerate > b_feerate;
+        }
+        return &(*a) > &(*b);
+    }
+};
+
+void MiniMiner::DeleteAncestorPackage(const std::set<MockEntryMap::iterator, IteratorComparator>& ancestors)
+{
+    for (const auto& anc : ancestors) {
+        auto vec_it = std::find(entries.begin(), entries.end(), anc);
+        assert(vec_it != entries.end());
+        entries.erase(vec_it);
+        entries_by_txid.erase(anc);
+    }
+}
+
+void MiniMiner::BuildMockTemplate(const CFeeRate& target_feerate)
+{
+    while (!entries_by_txid.empty()) {
+        // Sort again, since transaction removal may change some entries' ancestor feerates.
+        std::sort(entries.begin(), entries.end(), AncestorFeerateComparator());
+
+        // Pick highest ancestor feerate entry.
+        auto best_iter = entries.begin();
+        assert(best_iter != entries.end());
+        const auto ancestor_package_size = (*best_iter)->second.GetSizeWithAncestors();
+        const auto ancestor_package_fee = (*best_iter)->second.GetModFeesWithAncestors();
+        // Stop here. Everything that didn't "make it into the block" has bumpfee.
+        if (best_iter == entries.end() || ancestor_package_fee < target_feerate.GetFee(ancestor_package_size)) {
+            break;
+        }
+
+        // Calculate ancestors on the fly. This lookup should be fairly cheap, and ancestor sets
+        // change at every iteration, so this is more efficient than maintaining a cache.
+        std::set<MockEntryMap::iterator, IteratorComparator> ancestors;
+        std::set<MockEntryMap::iterator, IteratorComparator> to_process;
+        ancestors.insert(*best_iter);
+        to_process.insert(*best_iter);
+        while (!to_process.empty()) {
+            auto iter = to_process.begin();
+            assert(iter != to_process.end());
+            const CTransaction& tx = (*iter)->second.GetTx();
+            for (const auto& input : tx.vin) {
+                if (auto parent_it{entries_by_txid.find(input.prevout.hash)}; parent_it != entries_by_txid.end()) {
+                    to_process.insert(parent_it);
+                    ancestors.insert(parent_it);
+                }
+            }
+            to_process.erase(iter);
+        }
+        assert(ancestor_package_size == std::accumulate(ancestors.cbegin(), ancestors.cend(), 0,
+            [](int64_t sum, const auto it) {return sum + it->second.GetTxSize();}));
+        assert(ancestor_package_fee == std::accumulate(ancestors.cbegin(), ancestors.cend(), 0,
+            [](CAmount sum, const auto it) {return sum + it->second.GetModifiedFee();}));
+
+        // "Mine" all transactions in this ancestor set.
+        for (const auto& anc : ancestors) {
+            in_block.insert(anc->second.GetTx().GetHash());
```

Comment:
```suggestion
            in_block.insert(anc->first);
```
(functionally the same, just simpler)